Remove actor mode setting in favor of capability-based actors

pfefferle opened the pull request:
This commit removes the 'actor mode' option and all related code, migrating to a system where blog actors are always enabled and user actors are controlled solely via the 'activitypub' capability. Migration logic ensures sites previously in blog-only mode do not grant the capability to new users by default. All UI, constants, and logic related to actor mode selection have been removed or refactored, and related settings fields and filters are cleaned up for clarity and maintainability.

<!--- Provide a general summary of your changes in the Title above -->

Fixes #2418 

## Proposed changes:
- Deprecated actor mode constants and settings, replacing them with capability-based control
- Updated tests to reflect that both user and blog actors are always available
- Added migration to preserve blog-only mode behavior via a new flag
- Removed UI for actor mode selection

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [X] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [X] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Simplified actor management by removing the actor mode setting—blog actors are now always enabled, and user actors are controlled through capabilities for clearer, more consistent behavior.

</details>

pfefferle committed:
Remove actor mode setting in favor of capability-based actors

This commit removes the 'actor mode' option and all related code, migrating to a system where blog actors are always enabled and user actors are controlled solely via the 'activitypub' capability. Migration logic ensures sites previously in blog-only mode do not grant the capability to new users by default. All UI, constants, and logic related to actor mode selection have been removed or refactored, and related settings fields and filters are cleaned up for clarity and maintainability.

pfefferle committed:
Refactor tests to use string values for actor modes

Replaces usage of ACTIVITYPUB_ACTOR_MODE, ACTIVITYPUB_BLOG_MODE, and ACTIVITYPUB_ACTOR_AND_BLOG_MODE constants with their string equivalents ('actor', 'blog', 'actor_blog') throughout PHPUnit tests. Updates test logic and data providers to match the new string-based configuration, and removes unnecessary option updates where blog actor is now always available.

pfefferle committed:
Remove unused 'Profiles' settings section

Deleted the registration of the 'activitypub_profiles' settings section from the settings fields as it is no longer needed.

obenland commented on includes/collection/class-actors.php:
```diff
@@ -104,19 +103,11 @@ public static function get_by_username( $username ) {
 	 */
 	public static function get_id_by_username( $username ) {
-		// Check for blog user.
+		// Check for blog user (always enabled).
```
```suggestion
		// Check for blog user.
```
Not super helpful.

obenland commented on includes/collection/class-replies.php:
```diff
@@ -171,10 +170,7 @@ public static function get_context_collection( $post_id ) {
-				return false;
-			}
-
+			// Fallback to blog actor (always enabled).
```
```suggestion
			// Fallback to blog actor.
```

obenland commented on includes/scheduler/class-actor.php:
```diff
@@ -23,31 +22,24 @@ class Actor {
-
-		\add_action( 'add_option_activitypub_actor_mode', array( self::class, 'blog_user_update' ) );
-		\add_action( 'update_option_activitypub_actor_mode', array( self::class, 'blog_user_update' ) );
+		// Profile updates for blog options (always enabled).
```
```suggestion
		// Profile updates for blog options.
```

obenland commented on includes/scheduler/class-actor.php:
```diff
@@ -23,31 +22,24 @@ class Actor {
+		\add_filter( 'pre_set_theme_mod_custom_logo', array( self::class, 'blog_user_update' ) );
+		\add_filter( 'pre_set_theme_mod_header_image', array( self::class, 'blog_user_update' ) );
+
+		// Profile updates for user options (always enabled).
```
```suggestion
		// Profile updates for user options.
```

obenland commented on includes/transformer/class-post.php:
```diff
@@ -412,14 +412,9 @@ protected function get_type() {
-		return null;
+		// Blog actor is always enabled, so posts always have the blog as audience.
+		$blog = new Blog();
+		return $blog->get_id();
```
```suggestion
		// Posts always have the blog as audience.
		return ( new Blog() )->get_id();
```

obenland commented on includes/transformer/class-post.php:
```diff
@@ -1006,18 +1001,10 @@ private function get_public_interaction_policy() {
-			default:
-				return $this->get_actor_object()->get_id();
-		}
+		// Both user and blog actors are always enabled.
```
```suggestion
```

obenland commented on includes/wp-admin/import/class-starter-kit.php:
```diff
@@ -304,10 +303,7 @@ public static function import_options() {
-			return;
-		}
-
+		// Blog user is always available.
```
```suggestion
```

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -942,12 +940,11 @@ function activitypub_open_help_tab(event) {
-		if ( ! is_user_type_disabled( 'blog' ) ) {
-			\wp_add_dashboard_widget( 'activitypub_blog_profile', \__( 'ActivityPub Blog profile', 'activitypub' ), array( self::class, 'blogprofile_dashboard_widget' ) );
-		}
+		// Blog profile is always available.
```
```suggestion
```

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -395,22 +383,21 @@ public static function settings_page() {
-				'label'    => __( 'Followers', 'activitypub' ),
-				'template' => ACTIVITYPUB_PLUGIN_DIR . 'templates/followers-list.php',
-			);
+		// Blog profile and followers tabs are always available.
```
```suggestion
```

obenland commented on includes/constants.php:
```diff
@@ -63,11 +63,6 @@
-// Define Actor-Modes for the plugin.
-define( 'ACTIVITYPUB_ACTOR_MODE', 'actor' );
-define( 'ACTIVITYPUB_BLOG_MODE', 'blog' );
-define( 'ACTIVITYPUB_ACTOR_AND_BLOG_MODE', 'actor_blog' );
```
We should check if we break anything by removing these.

obenland commented on includes/functions.php:
```diff
@@ -368,7 +368,7 @@ function user_can_activitypub( $user_id ) {
 
 		case Actors::BLOG_USER_ID:
-			$enabled = ! is_user_type_disabled( 'blog' );
+			$enabled = true; // Blog user is always enabled.
```
Can we combine that with the application case?

obenland commented on includes/functions.php:
```diff
@@ -400,64 +396,30 @@ function user_can_activitypub( $user_id ) {
  * This function is used to check if the 'blog' or 'user'
  * type is disabled for ActivityPub.
  *
+ * Note: As of version 4.6.0, blog actors are always enabled and user actors
```
4.6.0 seems off.

Let's add a deprecation notice to this function?

pfefferle committed:
Update tests for removal of actor mode and blog actor fallback

Refactor and update PHPUnit tests to reflect the removal of the 'actor mode' option. Tests now expect invalid or missing user actors to always fall back to the blog actor, and assertions have been updated accordingly. Deprecated or removed code paths related to actor mode have been eliminated from test data providers and test logic.

pfefferle committed:
Remove actor mode logic and update related tests

Deprecated and simplified the Followers::maybe_add_inboxes_of_blog_user method to always return false, reflecting the removal of actor mode. Updated scheduler, outbox, dispatcher, query, and transformer tests to align with the new behavior, ensuring both blog and user actors are always available and adjusting assertions and helper methods accordingly.

pfefferle committed:
Remove references to blog actor always being enabled

Updated comments, documentation, and test descriptions to remove or rephrase statements about the blog actor always being enabled. This clarifies that blog actors are available, but avoids implying a special always-on status, aligning with recent changes to actor mode and capability handling.

pfefferle committed:
Remove 'always enabled' comments from actor-related code

Updated comments in actor-related classes and functions to remove references to 'always enabled' for blog and user options. This clarifies the current behavior and aligns documentation with recent changes to actor capability handling.

pfefferle committed:
Refactor actor mode to use constants and minor cleanups

Replaced hardcoded actor mode strings with defined constants for better maintainability and backward compatibility. Added actor mode constants in constants.php, updated migration logic to use these constants, and made minor code cleanups in functions.php and class-post.php.

pfefferle committed:
Remove deprecated maybe_add_inboxes_of_blog_user method and tests

Deleted the deprecated Followers::maybe_add_inboxes_of_blog_user method and its related PHPUnit tests, as actor mode has been removed. Also updated comments and minor code style in transformer test and added a phpcs ignore in outbox test.

pfefferle committed:
Update deprecated actor mode comment and improve test

Clarified the deprecation comment for actor mode constants in constants.php. Updated the Webfinger integration test to dynamically check the blog actor subject using the site's host instead of a hardcoded 'localhost' value.

pfefferle commented on includes/scheduler/class-post.php:
```diff
@@ -31,12 +31,12 @@ public static function init() {
+	 * @param \WP_Post|null $post_before Post object before the update.
 	 */
-	public static function schedule_post_activity( $post_id, $post, $update, $post_before ) {
+	public static function schedule_post_activity( $post_id, $post, $update, $post_before = null ) {
```
@copilot open a new pull request to apply changes based on [this feedback](https://github.com/Automattic/wordpress-activitypub/pull/2420#discussion_r2495554295)

pfefferle committed:
Remove single user mode and related logic

Eliminated the is_single_user() function and all associated checks throughout the codebase, as actor mode has been removed. Updated related logic in comment handling, user counting, blog type determination, and transformers to reflect this change.

pfefferle committed:
Remove activitypub capability from users in blog mode

When switching to blog-only mode, this update removes the 'activitypub' capability from all users who previously had it, ensuring that user capabilities are consistent with the new mode.

pfefferle committed:
Update user query to use capability__in for activitypub

Replaces the meta_key/meta_value query with 'capability__in' => array('activitypub') when removing the activitypub capability from users. This improves query performance and aligns with WordPress best practices for capability checks.

obenland commented on includes/class-comment.php:
```diff
@@ -139,11 +137,6 @@ public static function are_comments_allowed( $comment ) {
-		if ( is_single_user() && \user_can( $current_user, 'publish_posts' ) ) {
-			// On a single user site, comments by users with the `publish_posts` capability will be federated as the blog user.
-			$current_user = Actors::BLOG_USER_ID;
-		}
```
> On a single user site, comments by users with the `publish_posts` capability will be federated as the blog user.

Will this not fail now?

pfefferle committed:
Remove single user mode checks from Blog and Opengraph

Eliminated logic and comments related to 'single user' mode in Blog::get_type() and Opengraph::add_metadata(). This simplifies the code by removing special handling for single user blogs.

obenland commented on includes/functions.php:
```diff
@@ -467,15 +426,12 @@ function is_user_type_disabled( $type ) {
-	) {
-		return true;
-	}
+	\_deprecated_function( __FUNCTION__, 'unreleased', '' );
```
```suggestion
	\_deprecated_function( __FUNCTION__, 'unreleased' );
```

pfefferle committed:
functions.php aktualisieren

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Refactor Blog class to use properties for type and posting restriction

Moved the 'type' and 'posting_restricted_to_mods' values from methods to protected properties. Removed redundant methods and conditional checks, simplifying the class structure and improving clarity.

pfefferle committed:
Mark filter as deprecated in docblock

Updated the docblock for the user type disabling filter to explicitly mark it as deprecated and recommend using capability management instead.

pfefferle commented:
I will work on a new version... These change are too destructive...
